manufacturers/EGM/models/eNose/uplink_decoder.py

```diff
@@ -0,0 +1,134 @@
+    return ((Param_byte & 0xf0) >> 4)*10 + (Param_byte & 0x0f)
+
+
+def ENose_Proto(payload_bytes):
+    expected_length = 24  #24 bytes are expected
+    if len(payload_bytes) < expected_length:
+        raise ValueError(f"The payload has an unexpected length: {len(payload_bytes)}. expected: {expected_length}.")
+
+    redundancy = read_n_r(payload_bytes[0])
+    timestamp_sec = read_timestamp(payload_bytes[1:5])
+    Ammonia = (payload_bytes[5] << 4) + ((payload_bytes[6] & 0xf0) >> 4)
+    Air_components = ((payload_bytes[6] & 0x0f) << 8) + payload_bytes[7]
+    Methane = (payload_bytes[8] << 4) + ((payload_bytes[9] & 0xf0) >> 4)
+    Air_contaminants = ((payload_bytes[9] & 0x0f) << 8) + payload_bytes[10]
+    Butane_Propane = (payload_bytes[11] << 4) + ((payload_bytes[12] & 0xf0) >> 4)
+    VOC_isobutane = ((payload_bytes[12] & 0x0f) << 8) + payload_bytes[13]
+    Hydrogen = (payload_bytes[14] << 4) + ((payload_bytes[15] & 0xf0) >> 4)
+    NO2 = ((payload_bytes[15] & 0x0f) << 8) + payload_bytes[16]
+    CO2 = (payload_bytes[17] << 4) + ((payload_bytes[18] & 0xf0) >> 4)
+    Temperature = ((((payload_bytes[18] & 0x0f) << 8) + payload_bytes[19])/10)-40
+    Pressure = ((payload_bytes[20] << 4) + ((payload_bytes[21] & 0xf0) >> 4))/100
+    Humidity = (((payload_bytes[21] & 0x0f) << 8) + payload_bytes[22])/10
+    voltage_V = round(read_battery(payload_bytes[23]),1)
+
+    eNose_json = json.dumps({'r':redundancy,
+                             'timestamp':to_date(timestamp_sec),
+                             'Ammonia':Ammonia,
+				             'Air_components':Air_components,
+				             'Methane':Methane,
+				             'Air_contaminants':Air_contaminants,
+				             'Butane_Propane':Butane_Propane,
+				             'VOC_isobutane':VOC_isobutane,
+				             'Hydrogen':Hydrogen,
+				             'NO2':NO2,
+				             'CO2':CO2,
+				             'Temperature':Temperature,
+				             'Pressure':Pressure,
+				             'Humidity':Humidity,
+				             'voltage':voltage_V},
+				             separators=(',', ':'))
+    return eNose_json
+
+def decode_payload(payload_string, port_string):
+    payload_bytes = bytes.fromhex(payload_string)
+    decoded_data_json = '{}'
+    if port_string == '1':
+        decoded_data_json = ENose_Proto(payload_bytes)
+  
+    return decoded_data_json
+
+def ngsild_instance(value, time, unitCode, dataset_suffix):
+    ngsild_instance = {
+        "type": "Property",
+        "value": value,
+        "observedAt": time
+    }
+    if unitCode is not None:
+        ngsild_instance['unitCode'] = unitCode
+    if dataset_suffix is not None:
+        ngsild_instance['datasetId'] = f"urn:ngsi-ld:Dataset:{dataset_suffix}"
+    return ngsild_instance 
+
+def ngsild_wrapper(input):
+    ngsild_payload = {}
+
+    # Device infos
+    if 'Ammonia' in input:
+        ngsild_payload['Ammonia'] = ngsild_instance(input['Ammonia'], input['timestamp'], 'raw', None)
```

Comment:
naming convention for NGSI-LD is camelCase so you should name attributes like this:
-> ngsild_payload['ammonia'] = ...
->ngsild_payload['airComponents'] =
-> ngsild_payload['vocIsobutane'] =
-> ngsild_payload['no2'] =

The 3rd parameter is the unitCode -> see: https://unece.org/fileadmin/DAM/cefact/recommendations/rec20/rec20_rev3_Annex3e.pdf
So for example for temperature it's "CEL", humidity is "P1", ... If no unit or unit is not found just leave none

On the 4th parameters, add "Raw" (upper case R) for all attributes